ni_measurementlink_service/session_management.py

```diff
@@ -138,6 +138,20 @@ def unreserve(self):
         """Unreserve sessions."""
         self._session_manager._unreserve_sessions(self._session_info)
 
+    def get_pin_based_session_info(self, pin: str) -> List[Tuple[SessionInformation, str]]:
+        """Returns the session and channel information based on the given pin name."""
+        pin_based_session_info = []
+        for session_info in self.session_info:
+            for channel_mapping in session_info.channel_mappings:
+                if pin == channel_mapping.pin_or_relay_name:
+                    pin_based_session_info.append((session_info, channel_mapping.pin_or_relay_name))
+                    continue
+
+        if len(pin_based_session_info) == 0:
+            raise KeyError(f"Pin {pin} not found")
+
+        return pin_based_session_info
```

Comment:
I believe this API is supposed to be used after reservation and before calling the `_create_nidcpower_session` api.
The create session api would accept this session information to either create a new session or attach to an existing one.

Reply:
I disagree. Using it that way makes the measurement slow for no reason. Measurements that support N homogenous sessions should manage a list of sessions, not create one session at a time.

Reply:
So is the suggestion here to return the actual sessions? This would make the API driver specific, and it would be hard for us to manage it if the pin names correspond to different driver sessions.

Reply:
Also, `_nifgen_helpers.py` and `_niswitch_helpers.py` could provide multi-session helper functions that create a list of sessions and yield it to the caller.

```
def create_sessions(
    session_infos: Sequence[nims.session_management.SessionInformation],
    session_grpc_channel: grpc.Channel,
    initialization_behavior: Optional[
        nidcpower.SessionInitializationBehavior
    ] = nidcpower.SessionInitializationBehavior.AUTO,
) -> Generator[List[nidcpower.Session], None, None:
    with ExitStack() as stack:
        sessions = [
            stack.enter_context(create_session(session_info, session_grpc_channel, initialization_behavior))
            for session_info in session_infos
        ]
        yield sessions
```

Reply:
okay, got it. @bkeryan 
Currently the create sessions API is just taking in one session information to return 1 session in python.
So,
1.  This should be changed or a new version of this API should be added that accepts mutliple session information and returns multiple sessions.
2.  There should be a generic set of APIs that return the session based on pin information, one for single and one for multiple sessions. And these APIs will take in sessions array, pin name and session information array as input and return the appropriate session(s) based on the inputs.
3. The create method should open the context of the session internally and yield it so that once the session is used it automatically is destroyed. But one question here is that, won't the yield statement wait until previously yielded session is used/closed. Will this be a problem?.

Reply:
> Currently the create sessions API

There is no create sessions API. There is only a create session API. One is singular and the other is plural.

> This should be changed or a new version of this API should be added that accepts mutliple session information and returns multiple sessions.

I am suggesting that you add a multi-session create sessions API for drivers where it is appropriate (nifgen and niswitch, to start with).

It should be a separate API that delegates to the single-session create session API.

> The create method should open the context of the session internally and yield it so that once the session is used it automatically is destroyed. But one question here is that, won't the yield statement wait until previously yielded session is used/closed. Will this be a problem?.

The create session API (singular) should not yield. It returns a session, which is a context manager, so `with create_session(...) as session:` already works fine.

I am suggesting that the create sessions API (plural) yield a list of sessions. I am not suggesting that it yield each individual session. `with create_sessions(...) as sessions:` will execute `create_sessions` up to the `yield` statement, then bind the yielded list to the sessions variable and execute the rest of the measurement until it exits the `with` block, and then it will execute the rest of `create_sessions` after the `yield` statement.

However, on further thought, this has a different problem: if you expect `create_sessions()` to return a list of sessions, you will be sorely disappointed. You have to use it in a `with` block or iterate through it with `for` or `next()`. So maybe there is a better way than what I have proposed.

> There should be a generic set of APIs that return the session based on pin information, one for single and one for multiple sessions. And these APIs will take in sessions array, pin name and session information array as input and return the appropriate session(s) based on the inputs.

Return the appropiate session(s) **and channel(s)**. There may be multiple channels per session.

I'm not committed to having this API take a sessions array. As I have previously suggested, it could return a session **index** instead, like TSM does. Look up `SemiconductorModuleContext.cs` in ni-central to see what I mean. Note that TSM's .NET API uses method overloading, so `GetNIDCPowerSessions()` returns a different type of pin query context depending on whether you pass in one pin or multiple pins.

I would really like to see this API used to **simplify** some examples (as opposed to making them more complicated or changing them to demonstrate a more complicated use case).

Reply:
I have updated the new APIs to return the session index and the channel list.
The changes on starting the context in the create sessions method is not yet done, the context is still entered in the measurement logic.
The changes in dcpower example has been reverted and niscope and nifgen examples code has been updated to demonstrate the new APIs